File: 0_7_Previsit_Postvisit.py
import threading


class Graph:
    def __init__(self, vertex):
        # number of vertices
        self.vertices = vertex
        # adjList for the Graph
        self.adjList = [[] for x in range(vertex)]
        # visited
        self.visited = [False for x in range(vertex)]
        # pre
        self.pre = [-1 for x in range(vertex)]
        # post
        self.post = [-1 for x in range(vertex)]
        # counter 
        self.counter = 1

    # add edges
    def addEdge(self, vertex1, vertex2):
        self.adjList[vertex1].append(vertex2)
        self.adjList[vertex2].append(vertex1)
    # assign preorder
    def preOrder(self, vertex):
        self.pre[vertex]=self.counter
        self.counter+=1

    # assign postorder
    def postOrder(self, vertex):
        self.post[vertex]=self.counter
        self.counter+=1

    # explore 
    def Explore(self, vertex):
        # mark visited for this vertex as true
        self.visited[vertex]=True

        # add counter to preOrder
        self.preOrder(vertex)

        # now check it's neighbour
        for ver in self.adjList[vertex]:
            if self.visited[ver]==False:
                self.Explore(ver)

        # counter to the postOrder
        self.postOrder(vertex)

    
    # DFS
    def DFS(self):
        for vertex in range(self.vertices):
            if self.visited[vertex]==False:
                self.Explore(vertex)

def main():
    vertex, edges = [int(x) for x in input('vertex and edges --> \t').split()]

    graph = Graph(vertex)
    for x in range(edges):
        vertex1, vertex2 = [int(x) for x in input().split()]

        graph.addEdge(vertex1, vertex2)

    graph.DFS()

    print("pre-->\t", *graph.pre)
    print("post-->\t", *graph.post)

threading.Thread(target=main).start()

'''
test case
input
5 5
0 1
0 2
0 4
1 3
1 4

output
pre-->   1 2 8 3 5
post-->  10 7 9 4 6

'''

File: 0_7_Previsit_Postvisit.java
import java.util.*;
public class Previsitpostvisit {
    // vertices in the graph
    private int vertices;
    // adjacent list
    private List<Integer> adjList[];
    // visited
    private boolean visited[];
    // pre
    private int pre[];
    // post
    private int post[];
    // cuunter
    private int counter=1;

    Previsitpostvisit(int v){
        vertices = v;
        adjList = new ArrayList[v];// assign length
        visited = new boolean[v];
        pre = new int[v];
        post = new int[v];

        for(int x=0; x<v; x++){
            adjList[x]=new ArrayList();
            visited[x] = false;
            pre[x]=-1;
            post[x]=-1;
        }
    }

    // add edges
    void addEdge(int vertex1, int vertex2){
        adjList[vertex1].add(vertex2);
        adjList[vertex2].add(vertex1);
    }

    // preorder assignment
    void preOrder(int vertex){
        pre[vertex]=counter;
        counter+=1;
    }

    // postOrder
    void postOrder(int vertex){
        post[vertex]=counter;
        counter+=1;
    }
    
    // explore
    void Explore(int vertex){
        // first mark as true of this index
        visited[vertex]=true;
        // assign for the preorder
        preOrder(vertex);

        // now iterate through the all neighbour of this vertex
        Iterator<Integer> i = adjList[vertex].listIterator();
        while(i.hasNext()){
            int ver = i.next();
            if (visited[ver]==false){
                Explore(ver);
            }
        }
        // for the postOrder
        postOrder(vertex);
    }

    //dfs
    void dfs(){
        for(int vertex=0; vertex<vertices; vertex++){
            if (visited[vertex]==false){
                Explore(vertex);
            }
        }
    }

    public static void main(String[] args){
        Scanner scanner = new Scanner(System.in);
        int vertex = scanner.nextInt();
        int edge = scanner.nextInt();

        // make the object of the class
        Previsitpostvisit graph = new Previsitpostvisit(vertex);

        // make the adjacency list
        for (int x=0; x< edge; x++){
            int vertex1 = scanner.nextInt();
            int vertex2 = scanner.nextInt();

            graph.addEdge(vertex1, vertex2);
        }
        scanner.close();

        // do dfs
        graph.dfs();
        
        System.out.print("pre --> \t");
        for(int x=0; x<vertex;x++)System.out.print(graph.pre[x]+ " ");
        System.out.println();
        System.out.print("post --> \t");
        for(int x=0; x<vertex;x++)System.out.print(graph.post[x]+ " ");

    }
}
